aten/src/ATen/core/dispatch/Dispatcher.cpp

```diff
@@ -498,24 +498,38 @@ std::vector<OperatorName> Dispatcher::getRegistrationsForDispatchKey(c10::option
   });
 }
 
-int64_t Dispatcher::sequenceNumberForRunningRecordFunction(DispatchKey dispatchKey) {
+int64_t Dispatcher::sequenceNumberForRunningRecordFunction(DispatchKey dispatchKey, DispatchKeySet dispatchKeySet) {
   int64_t seq_num = -1;
   // Setting sequence number in the Autograd case to associate
   // the forward range with the corresponding Autograd's node
-  if (isIncludedInAlias(dispatchKey, DispatchKey::Autograd) && at::GradMode::is_enabled()) {
+
+  // Note: this records a sequence number for both Autograd keys, and for
+  // non-Autograd keys where the dispatchKeySet still contains an autograd key.
+  // This means that we might collect the same sequence nubmer two different
+  // events if they all occurred above Autograd and still had the Autograd
+  // dispatch key in the dispatch key set.
+  // However, this usually doesn't happen: normally the first call will
+  // go through the call() or callBoxed() path in the dispatcher, while
+  // subsequent redispatches go through redispatch() or redispatchBoxed().
+  // `call` has profiler instrumentation, whereas `redispatch` doesn't.
+  // So usually, we'll collect a sequence number on the first call() if the
+  // dispatch keys contain autograd, and not on subsequent redispatches.
+  bool dispatchHasAutograd = !(dispatchKeySet & autograd_dispatch_keyset).empty();
```

Comment:
btw my suggestion was to check the exclude set here.
The way autograd works is that the key is added to every single Tensor. So everything will always dispatch to autograd unless:
- Autograd key is in the exclude set
- Inference mode is running

Since inference mode implies no_grad mode (checked below), I think it is enough to check the tls exclude set here. That way you don't need to pass any argument in and can simplify the overall logic quite a bit.

Reply:
ah I see, thanks! I figured we didn't need to check given we checked no_grad mode below, but your explanation makes it clear why checking the exclude set is better

Reply:
@albanD - actually, I think the exclude set should already be handled:

from callBoxed, this gets the dispatch key set, which is eventually passed into the sequenceNumberForRunningRecordFunction function that I'm modifying in this PR:
https://github.com/pytorch/pytorch/blob/56dd7603da0d9f8b7f7439cf2091a1fd306983f5/aten/src/ATen/core/dispatch/Dispatcher.h#L718

and getDispatchKeySetBoxed eventually calls computeDispatchKeySet:
https://github.com/pytorch/pytorch/blob/56dd7603da0d9f8b7f7439cf2091a1fd306983f5/aten/src/ATen/core/dispatch/DispatchKeyExtractor.h#L170

computeDispatchKeySet checks the exclude set:
https://github.com/pytorch/pytorch/blob/56dd7603da0d9f8b7f7439cf2091a1fd306983f5/aten/src/ATen/core/dispatch/DispatchKeyExtractor.h#L46

Based on this, do you think it's fine to leave it as is? (Is GradMode expected to show the exact same info as the include/exclude dispatch key sets? I checked and it wasn't obvious to me)

Reply:
gradmode and include/exclude can diverge for example (if this is what you are asking) because when you redispatch in the autograd kernel, you first disable autograd via the exclude key. If you later have a callBoxed  in torch dispatch, gradmode can be still enabled, but the autograd key has now been excluded.

Reply:
> actually, I think the exclude set should already be handle

I think the idea is that if checking only exclude set is sufficient to determine whether autograd kernel is reached, then we can avoid threading dispatch key at all these callsites and instead just check TLS inside sequenceNumberForRunningRecordFunction

However, if you create tensors within inference mode and then use them outside of inference mode, you will not reach the autograd kernel even though you are outside of no_grad, and autograd is not in TLS exclude set.

So, TLDR - the current appraoch is probably fine.